flux-middle/src/rty/subst.rs

```diff
@@ -332,7 +333,14 @@ impl TypeFolder for GenericsSubstFolder<'_> {
     fn fold_ty(&mut self, ty: &Ty) -> Ty {
         match ty.kind() {
             TyKind::Param(param_ty) => self.ty_for_param(*param_ty),
-            TyKind::Indexed(BaseTy::Param(param_ty), idx) => self.bty_for_param(*param_ty, idx),
+            TyKind::Indexed(BaseTy::Param(param_ty), idx) => {
+                let idx = idx.try_fold_with(self).into_ok();
+                if self.generics.is_some() {
```

Comment:
Yes, this was a bit delicate. 

I had a failing test `refined_type_var00` because of the following:

```rust
#[flux::sig(fn<T as base>(b: bool, x: T[@n], y: T[@m]) -> T[if b { n } else { m }])]
pub fn choose<T>(b: bool, x: T, y: T) -> T {
    if b {
        x
    } else {
        y
    }
}

pub fn test01() {
    assert(choose(true, 0, 1) == 0);
    assert(choose(false, 0, 1) == 1);
}
```

Now, at the callsite `choose(true, 0, 1)` there are *two* substitutions going on in the signature for `choose` i.e. for the `T[@n]` and `T[@m]`

1. `T` -> `i32{v: ...}`
2. earlybound `n, m` -> fresh-evars.

The trouble is that if you solely rely on the `bty_for_param` code below, then the `n, m` substitutions "get lost" and so those variables are not "solved for" and you get that pesky instantiation error. So instead, we _first_ substitute for the indices, and then let `bty_for_param` do its business.

Maybe I should add this as a comment to the code?

LMK if that makes sense.

Reply:
ah yeah, this is delicate. although I still don't follow. the problem you describe sounds related to not calling `fold_with` on the `idx` (which was a bug before). My question is why check for `is_some` here and then again in `bty_for_param`, I feel we can write it like

```rust
let idx = idx.fold_with(folder);
self.bty_for_param(*param_ty, &idx)
```

Reply:
duh! of course, fixed.